pkg/crawl/class_hash.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2021 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crawl
+
+import (
+	"bytes"
+	"crypto/md5"
+	"fmt"
+	"io"
+
+	"github.com/palantir/log4j-sniffer/pkg/java"
+)
+
+const maxClassSize = 0xffff
+
+var classByteBuf bytes.Buffer = bytes.Buffer{}
```

Comment:
Is there a reason that this is kept as an instance variable rather than a variable within the function? It means that `LookForHashMatch` can't be run in parallel, and not clear to me that the buffer allocation itself is that expensive (although may not matter).

Reply:
It was an attempt to keep memory usage type when running in the default configuration - we currently manage to run with a memory budget of 10-20mb across our production fleet and I was (perhaps unreasonably) concerned that not reusing this buffer would lead to spikes of memory use prior to GC.

It definitely does prevent parallelism but right now are actually add rate limiting to try and slow things down. I think this is acceptable for now but if we wanted to make an option that tried to use all possible resources we'd need to revisit this.

Reply:
OK mainly just surprised that the allocation/deallocation would show up at all, but not a big deal in this context -- fine to leave as-is.